### ISSUE TITLE
RFC: quo vadis, `xp_assert_*` infrastructure?

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#21044]
This topic sits at the interface of at least three-four different issues, at least two of which I seem to have a knack of running into. I don't consider myself the right (much less the best) person to write about this, but here we are...

Those four issues are:
1. [Array API](https://data-apis.org/array-api/latest/) standard adoption
1. Treatment of 0d-arrays in NumPy & SciPy
1. General testing facilities
1. Desirable fidelity in testing

I won't write about the first topic, which is covered elsewhere (e.g. https://github.com/scipy/scipy/issues/18286), I just take it as an obviously beneficial goal and acknowledge that its implementation is happening already.

## Treatment of 0d-arrays

NumPy arrays are nicely generic over their dimension, which can range from 0 to `NPY_MAXDIMS` (==32). NumPy also has its own scalars to complement the Python-native ones (e.g. `np.int64(...)` is obviously limited in range but has constant footprint, whereas `int(...)` is unbounded in both value and footprint). One long-standing historical inaccuracy has been that one only needs to breathe at a 0d-array the wrong way, and it decays into a scalar:
```python
>>> x= np.asarray(1.)
>>> -x
np.float64(-1.0)
```

This is BAD from a number of perspectives, principally those two objects have different APIs (although many methods/fields actually are compatible). Another issue is typing, e.g. it's ~impossible~ very hard to write a function
```
func(x: np.array) -> np.array
```
because there are so many ways that 0d inputs might actually return scalars, forcing the return type to be `np.array | scalar`.

For these reasons (and others), all array standard API implementations except NumPy actually keep 0d arrays as 0d arrays. Great! However, SciPy is still _very_ closely related to NumPy, and we cannot really get away from the reality of 20+ years of legacy that has grown around this. [NEP 56](https://numpy.org/neps/nep-0056-array-api-main-namespace.html) even call this out specifically (quoted):

> [!NOTE]
> We note that one NumPy-specific behavior that remains is returning array scalars rather than 0-D arrays in most cases where the standard, and other array libraries, return 0-D arrays (e.g., indexing and reductions). Array scalars basically duck type 0-D arrays, which is allowed by the standard (it doesn’t mandate that there is only one array type, nor contains `isinstance` checks or other semantics that won’t work with array scalars). There have been multiple discussions over the past year about the feasibility of removing array scalars from NumPy, or at least no longer returning them by default. However, this would be a large effort with some uncertainty about technical risks and impact of the change, and no one has taken it on. Given that array scalars implement a largely array-compatible interface, this doesn’t seem like the highest-prio item regarding array API standard compatibility (or in general).

It's a great pity that this change ended up not making NumPy 2.0, but such is life. That said, everybody seems to agree that this clean-up would be worthwhile, just that the effort involved would possibly be very large. In any case, there is interest in both [NumPy](https://github.com/numpy/numpy/issues/24897) and [SciPy](https://github.com/scipy/scipy/issues/18768).

However, the reality is the vast majority of code in SciPy currently only uses NumPy and is susceptible to this. The `stats` module is doubly affected, for one because it's where some of the first bits of array API work are landing, and also because it intrinsically has a large number of operations that do dimension reduction (like the mean), which is even more likely to produce scalars than something that just keeps the same shape.

Additionally, there were a whole host of other inconsistencies, so the pragmatic decision of the stats team was to introduce consistency in the only way available that doesn't require fighting against the NumPy tide, i.e. by declaring that all their returns be scalars rather than 0d arrays. This is directly counter to the above-mentioned [RFC](https://github.com/scipy/scipy/issues/18768), but after having talked with @mdhaber about this, the long-term vision is still the same - only that it would avoid piecemeal change, but rather do one big change over (probably SciPy 2.0 material) if/once NumPy is ready.

## General testing facilities & strategies

For the purpose of this discussion, I'm going to distinguish three levels of testing infrastructure, which uses they're aimed at, and which expectations they need to satisfy:

### 🥇 array-api standard defines its own testing functions

Coming into this topic, I had naïvely assumed that this was the case already. Given the discussion here, it's a good thing though that these things can still evolve quite freely while we figure out the needs of the API, as opposed to being already frozen in a hard-to-change standard.

Still, if some `assert_*` functions ever make it into the array api standard, they obviously need to cover a very wide range of use-cases (despite the dispatch to a given backend potentially being limited in various ways), and be "gold standard" level API.

### 🥈 SciPy-wide testing facility

One level down, not ecosystem-wide but "only" SciPy-wide, testing infrastructure like the `xp_array_*` functions must be at least as general/usable as `np.assert_*` (this is especially the case as we seek to replace the latter with the former in the drive to enable other array backends). In particular, they must satisfy at least:
```
# identity
xp_assert_equal(a, a)  # passes for all conceivable kinds of inputs (at least those that np.assert_* manages)
# commutativity
xp_assert_equal(a, b)  # passes if and only if...
xp_assert_equal(b, a)  # ... also passes
```
and generally, satisfy a very high standard of implementation/documentation to support general use in SciPy

### 🥉 private testing facility

In this case, there are no real requirements, every module can do what's reasonable (without gratuitous duplication). It should therefore be restricted to cases where a general solution is not possible/feasible.

## Desirable fidelity in testing

I'm making this a separate point because some parts of this keep being rediscussed in various places. There are several levels of fidelity one can aim for with comparing two python objects:
* I) `a.__eq__(b) == True`, i.e. whatever someone implemented in the class-specific `==` overload
* II) Relevant metadata of `a` and `b` matches as well (types, shapes, etc.)
* III) Identity of `a` and `b`

Identity is irrelevant for us (aside from handling `None` and `np.nan` in some code branches), so we are explicitly not talking about that -- everything has value semantics. 

However, II) is eminently important for a library, and even more so for something as fundamental as SciPy. We absolutely _do not_ want to be lax and loose about whether our functions return the expected shape/dtype etc. This was how it was like in the bad old days, and we're very actively trying to move away from that. I formulate this first and foremost as my own opinion, although I have discussed this with several other maintainers and am under the impression that they share this view. For the sake of balance, @ilayn [provides](https://github.com/scipy/scipy/issues/20959#issuecomment-2168832284) a counter-argument:

> This is the black hole javascript fell with `==` and `===`. I don't think we need to repeat their mistake. `assert_equal_###` is about the values being equal and not the metadata about things. `42 == 42.0` must return `True` regardless of the library.
> 
> If we need to enforce dtype, shape _and_ the values we need to find another name for it in my opinion `assert_same_array_props` or whatever.

I strongly disagree on this front. The default testing infrastructure IMO must use the strictest reasonable defaults, especially if we now have the chance and willingness to clean up a lot of the old inaccuracies through the array API effort. Opting out of that fidelity (e.g. where it's unimportant or too difficult to achieve) should be explicit. The current `xp_assert_*` functions already provide several switches to opt out of certain checks:
https://github.com/scipy/scipy/blob/f7e0d1a0d83741ea8efebf33f5ccbda12d1d8503/scipy/_lib/_array_api.py#L312-L313

With the exception of `allow_0d` (see below), I think this is good design. Unsurprisingly, conversions to the array API (and thus using `xp_assert_*`) forces a clean-up of currently fuzzy test expectations. This is IMO a good thing, and it can (and already does) happen incrementally as functions are converted.

## Current Situation

The problem is that the current `xp_assert_*` functions, both by location (`scipy/_lib/_array_api.py`) and by naming, look like they're SciPy-wide or even ecosystem-wide, but in reality are so-far only considered "a private testing utility" by the people working on SciPy's array API implementation. This mismatch needs to be resolved -- either the current `xp_assert_*` functions become private to the array API effort (not desirable from my POV), or they need to be adapted to satisfy a wider set of use-cases.

Additionally, as mentioned in the 0d-discussion, `stats` (and other modules) want to increase the consistency of their functions. Given the realities of NumPy's 0d-treatment, their pragmatic choice was to ban all 0d-arrays from being returned by `stats` functions (at least those that have successfully undergone conversion to the array API).

The combination of these two things lead them to disallow 0d-arrays in `xp_assert_*` as the "actual" value (as opposed to "desired"; see signature above), unless `allow_0d=True` was specified. This leads to some very unintuitive results (c.f. #20959):
```
>>> from scipy._lib._array_api import xp_assert_close, xp_assert_equal
>>> a = np.array(42)
>>> b = a[()]
>>> xp_assert_equal(a, a)  # AssertionError
>>> xp_assert_equal(a, b)  # AssertionError
>>> xp_assert_equal(b, a)  # passes...
>>> xp_assert_equal(b, a, allow_0d=True)  # AssertionError, despite opting into apparently looser checks?!
```

This is simply not suitable for something SciPy-wide and must be fixed.

## Proposed Solution

One simple approach ([draft](https://github.com/scipy/scipy/pull/21026)) is to replace the default `allow_0d=False` with `check_0d=True`, with the following behaviour:
```
>>> a = np.array(42)
>>> xp_assert_equal(a, a)   # passes
>>> xp_assert_equal(a, 42)  # AssertionError
>>> xp_assert_equal(42, a)  # AssertionError
>>> xp_assert_equal(np.int64(42), a)  # AssertionError
>>> xp_assert_equal(a, np.int64(42))  # AssertionError
>>> xp_assert_equal(np.int64(42), a, check_0d=False)  # passes
>>> xp_assert_equal(a, np.int64(42), check_0d=False)  # passes
>>> xp_assert_equal(a, 42, check_0d=False)  # passes
>>> xp_assert_equal(42, a, check_0d=False)  # passes
```

However, in order for modules (like `stats`) that are especially affected by the numpy inconsistencies to opt out of having to set `check_0d=False` in a lot of places, @mdhaber proposed a factory-method to _create_ the three `xp_assert_{equal,close,less}` functions, with some test-module-specific configuration. This should perhaps also mitigate the concerns by @ilayn about potential overreach of this effort, as the choice only needs to be done once per module.

The wrinkle is that the four kwargs (under the above proposal) `check_{namespace,shape,dtype,0d}` are still not enough for the situation in which the stats modules (and others) finds itself, which still would like to prohibit 0d arrays completely. But by the factory-approach, that particular option could be contained to just the method creation.

**Comment by fancidev** [comment]
Nice write-up!

FWIW, I’d like to mention that the current `xp_assert_equal` implementation treats `+0.0` as being equal to `-0.0` under PyTorch, while being unequal under `numpy`. As there is no spec on what `xp_assert_equal` is supposed to do, this inconsistency in behavior is not “wrong”. It just means that the sign of zero cannot be tested with `xp_assert_equal`.

**Comment by h-vetinari** [comment]
Thanks!

> FWIW, I’d like to mention that the current `xp_assert_equal` implementation treats `+0.0` as being equal to `-0.0` under PyTorch, while being unequal under `numpy`.

Yeah, in this case, my tendency is the opposite: that distinction is really pointless IMO (and it has zero APIconsequences, in contrast to the other `check_*` questions).

**Comment by fancidev** [comment]
Well, in some cases the sign of zero matters. A concrete recent case is gh-20929, where an accurate inverse CDF function for the Cauchy distribution would involve the expression `cosPi(p) / sinPi(p)`, and the sign of `sinPi(1)` (zero) makes a difference of giving the right or wrong result. Imagine if one wants to test this across array backends.

**Comment by h-vetinari** [comment]
> Well, in some cases the sign of zero matters.

Negative zero is an abomination. It's a figment of the imagination of the IEEE floating points, because a bit pattern was free, and they didn't have anything better to do than allow it. In this particular case, `sinPi(-0.)` should be zero, which is also how numpy treats this question (correctly):
```
>>> np.sign(0.)
0.0
>>> np.sign(-0.)
0.0
```

For the factory, I was imagining a `forbid_0d`-kwarg, which would satisfy the `stats` use-case. I wouldn't be opposed in principle to having a `forbid_negative_zero` argument that we could use to get rid of negative zero in our function returns (even though the logistics of that are harder, because we'd have to introspect potentially large arrays per element), but I doubt it's a worthwhile effort, compared to all the other things we could be fixing.

> A related case is the propagation of NaN payload.

Likewise, we shouldn't care about those either (another floating point misfeature). We already have to distinguish between various different _types_ of missing values, like `np.Nan`, `np.NaT`, `None`, etc. Luckily, signalling NaNs basically don't occur in the Python ecosystem (AFAIK). We should keep it that way (and if necessary, force all floating point NaNs to be bitwise identical with `np.nan`).

> Other complicated scenarios may include testing array layout (C vs Fortran), which makes a difference in memory locality and hence performance, or sparse vs dense array

From my POV: the former should not be the default (again the standard I'm applying is: does it have API impact?), but it could be opt-in, perhaps. Sparse vs. dense would make sense to capture indeed.

**Comment by lucascolley** [comment]
Thanks @h-vetinari ! I agree with the vast majority of points here, will reply to everything I'm not so sure on.

> SciPy-wide, testing infrastructure like the xp_array_* functions must be at least as general/usable as np.assert_* (this is especially the case as we seek to replace the latter with the former in the drive to enable other array backends). In particular, they must satisfy at least ... identity ... commutativity

Okay, this position seems a bit more reasonable in the context of the assertions eventually becoming replacements for the `np.testing` versions. I still think it is a bit too harsh to say that these properties *must* be satisfied - for example, if people are running into issues with lack of commutativity (with any reasonable implementation of these assertions), they are almost certainly getting `actual` and `desired` mixed up. Perhaps it does make sense to say that you should be able to pass your desired result to `actual` and your actual result to `desired` in the default configuration (and I imagine this already happens in various places across the ecosystem), but I don't think we should too quickly dismiss the idea that `actual` and `desired` could really be distinguished (and usefully so), in the default configuration.

Self-identity, on the other hand, does seem very sensible for the default configuration.

>>If we need to enforce dtype, shape and the values we need to find another name for it in my opinion assert_same_array_props or whatever.
> 
> I strongly disagree on this front. The default testing infrastructure IMO must use the strictest reasonable defaults, especially if we now have the chance and willingness to clean up a lot of the old inaccuracies through the array API effort. Opting out of that fidelity (e.g. where it's unimportant or too difficult to achieve) should be explicit. The current xp_assert_* functions already provide several switches to opt out of certain checks

Completely agreed that "the default testing infrastructure ... must use the strictest reasonable defaults". However, if that breaks the mental model of a function with the word "equal" in its name too much, I see no fundamental issue in changing the name.

>  @mdhaber proposed a factory-method to create the three xp_assert_{equal,close,less} functions, with some test-module-specific configuration. This should perhaps also mitigate the concerns by @ilayn about potential overreach of this effort, as the choice only needs to be done once per module.

I like this idea.

> For the factory, I was imagining a forbid_0d-kwarg, which would satisfy the stats use-case.

It sounds like a `forbid_scalar` would also be useful in `ndimage`!

> sparse vs dense array

This should be caught by `check_namespace`, unless there are sparse and dense arrays being produced by the same namespace.

**Comment by ilayn** [comment]
Great post @h-vetinari ! I should kick myself off of the pedestal as the opposing party here and clarify the point I want to make, because I agree with all this, except the function name. These all are sensible and flexible defaults.

But the action we are performing is not an equality test but a strict comparison. that's what I mean by the javascript remark. And I want that strict comparison because it looks something I would start using right away based on your API suggestion. But the name 😊

We have spent significant amount of time and effort to get to a point where we use ‛assert_allclose‛ for float comparison as equality won't make sense and ‛assert_equal‛ for integral quantities regardless of the operand type. This also follows the underlying Python behavior as it should be the guiding mental model (as much as we can meaningfully accomodate obviously). As I had been subjected to fortran,c, c++ very recently,one of most jarring annoyances is unnecessary type casting of things just to perform the most mundane things. So Python gets thus just right not too loose (say bytes vs str) or too strict. 

Having said all this, i would expect an api perform equality if it says so on the tin. Meaning that we should not hijack the ‛assert_equal‛ api for doing more stuff but enhance the testing suite catalogue. In fact this is already covered with [‛assert_array_equal‛](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html). Then it is obvious that things should be both array and equal and your cases fail with proper mental expectations. In my opinion that should be the place to overload all this power and flexibilty.

This would leave the convenience of ‛assert_equal‛ alone while significant functionality is offered with ‛assert_array_equal‛. It is a shame that we had to disturb the naming suite for almost equal stuff because legacy reasons but nevertheless hope my message gets across with this example.

Otherwise we might fall victim to a feature creep that a once- pretty straightforward api gains a significant functionality forcefully tucked inside settings and mutating behavior (‛git‛ is of notorious fame)

**Comment by ilayn** [comment]
Just to mention one more time within this context, array api should provide this. So +1 to your yellow medal point.

**Comment by rgommers** [comment]
To echo the sentiments above: great write-up, thanks @h-vetinari! I agree with pretty much all the summary (it looks accurate) and most (all?) of your opinions. 

> The problem is that the current `xp_assert_*` functions, both by location (`scipy/_lib/_array_api.py`) and by naming, look like they're SciPy-wide or even ecosystem-wide, but in reality are so-far only considered "a private testing utility" by the people working on SciPy's array API implementation

I don't agree with this assessment/position. I certainly consider them SciPy-wide functions. They're already heavily used in a lot of files across a number of submodules. And I believe they were also designed that way, as a generalization of the equivalent numpy functions. The 0-D vs. scalar thing is honestly pretty niche, and does not detract from how I'd look at these functions.

> However, in order for modules (like `stats`) that are especially affected by the numpy inconsistencies 

So this is mostly about `stats` I think, right? With `special` the second one, but there almost all functions are implemented as ufuncs so they are guaranteed to have uniform behavior anyway, so it doesn't matter as much whether this is checked for each function.

> One simple approach (https://github.com/scipy/scipy/pull/21026) is to replace the default `allow_0d=False` with `check_0d=True`,

This looks quite reasonable to me, it's a logical change to make.

> by the numpy inconsistencies to opt out of having to set check_0d=False in a lot of places, @mdhaber proposed a factory-method to create the three `xp_assert_{equal,close,less}` functions, with some test-module-specific configuration ... are still not enough for the situation in which the stats modules (and others) finds itself, which still would like to prohibit 0d arrays completely.

This looks like a less then ideal choice, and I kinda read it as an over-reaction to a corner case consistency issue. These functions will be used a _lot_ throughout the whole code base. In the future, they may also be generalized into common functions that can be shared with other packages in the ecosystem. That means I think that they should be easy to understand. I can't be sure without seeing an implementation, but this sound like significant over-engineering that will make understanding and debugging harder.

My assessment of the 0-D vs. scalar is:
- it's a corner case, which we never consistently checked for in the 20+ years of scipy history
- similar to other corner cases like empty array input, or specific behavior like infs/nans
- would be good to make more consistent, but testing that can be done in a number of ways and does not need to be done for every invocation of a test function.
  - each function in our API is typically tested many times with different input values, keywords, etc. It would be fine to add a single function call with 0-D array input for each such function, and verify whether the output is a 0-D or scalar. (a la what @ev-br recently did for empty inputs - those are in the same boat, and much more important to test for, as evidenced by lots of bug reports). That way, the new keyword `allow_0d` proposed in gh-20959 would not even be needed. Which may be nicer for generality, since there won't be much demand for that keyword outside of SciPy. 
  - Also important for generality: the keyword is numpy-specific, since other libraries all don't have scalars. So it's not ideal to include it.
  - Note, `numpy.testing` also doesn't have this check, and numpy doesn't consistently test its own functions that way - it's never been high-enough prio for someone to care as far as I know.

**Comment by rgommers** [comment]
> Because, it seems to me that this is going to be the precursor to the prospective `scipy.testing` module.

@ilayn I wouldn't worry about this. I think no one wants to see that. We can agree right here that this is out of scope for SciPy, and it will stay that way. In case I ever change my mind in the future (highly unlikely), please feel free to point at this comment and veto my change of heart:)

**Comment by h-vetinari** [comment]
Thanks everyone for the inputs! I was hoping for @mdhaber to join in as well, especially now that there's a counter-argument to the factory-approach.

> > look like they're SciPy-wide or even ecosystem-wide, but in reality are so-far only considered "a private testing utility" by the people working on SciPy's array API implementation
> 
> I don't agree with this assessment/position. I certainly consider them SciPy-wide functions.

This was a verbatim quote, which I just didn't source to avoid pointing any fingers. My intention was to sort out this discrepancy, and I agree that they should be SciPy-wide.

> each function in our API is typically tested many times with different input values, keywords, etc. It would be fine to add a single function call with 0-D array input for each such function, and verify whether the output is a 0-D or scalar.

I'm not 100% on board with this, because it opens the door to only handling that distinction only under the very limited conditions of that single (usually trivial) test, rather than ensuring consistency across all call-paths. I think my experience w.r.t. overlaps with what @mdhaber et al. experienced, in that it's easy to become inconsistent again across a variety of code-paths returning earlier/later than others, and forgetting the "dimensional consistency" in cases that aren't covered by the smoketest.

> Also important for generality: the keyword is numpy-specific, since other libraries all don't have scalars. So it's not ideal to include it.

That's a good point, but maybe speaks again to the factory approach, as it's not something we'd want to have in the array api reference implementation?

> Note, `numpy.testing` also doesn't have this check

The API for the functions in `numpy.testing` already look very different. They have a global `strict=` opt-in, not the granular `check_<X>` that we have here for now.

**Comment by h-vetinari** [comment]
> numpy doesn't consistently test its own functions that way - it's never been high-enough prio for someone to care as far as I know.

I think this issue will continue to be painful until numpy fixes this inaccuracy. But would this even be a feasible change without waiting for `numpy 3.0` (perhaps the blast radius is small enough that it can be done with a user opt-in that becomes default after 1-2 releases...)?

**Comment by h-vetinari** [comment]
@mdhaber has told me privately that he has no capacity to deal with this subject ATM, but is counting on `stats` testing not regressing from the POV of avoiding 0d arrays.

@rgommers, do you have an idea for how to square that circle (if it's not the factory approach)? Or perhaps the factory approach is the least bad option for now (also allowing the xp-testing to stay independent of numpy-specifics)? Would also appreciate an opinion on the feasibility of cleaning this up in numpy 2.x (rather than 3.0). Thanks :)

**Comment by h-vetinari** [comment]
PS. Perhaps the least invasive approach for now is to go with https://github.com/scipy/scipy/pull/21026 as the default, but modify those functions as required only in `stats` (or wherever else it is needed), e.g.
```diff
-from scipy._lib._array_api import xp_assert_equal, xp_assert_close
+from scipy._lib._array_api_no_0d import xp_assert_equal, xp_assert_close
```
and then define the functions in `_array_api_no_0d` as a thin wrapper (that just forbids 0d-args in `actual`) around the core implementation in `_array_api`. That would avoid the complexity of the factory approach, but still make it a per-module opt-in to one or the other behaviour.

**Comment by h-vetinari** [comment]
@rgommers could you comment on the idea with `_array_api_no_0d` above?

**Comment by rgommers** [comment]
> and then define the functions in `_array_api_no_0d` as a thin wrapper (that just forbids 0d-args in `actual`) around the core implementation in `_array_api`. That would avoid the complexity of the factory approach, but still make it a per-module opt-in to one or the other behaviour.

This sounds like a good and clean solution that meets the need for `stats` strict checking while keeping the main implementation clean.

**Comment by h-vetinari** [comment]
The implementation for this RFC is now ready in #21026 (at least for review).

**Comment by lucascolley** [comment]
See gh-21413 for another attempt at an implementation here. Below I evaluate my design choices against the alternatives.

---

# Problem instances

## (1) Full module returning 0-D arrays

### Solution

```python
from scipy._lib._array_api import xp_assert_close
...
xp_assert_close(actual, desired) # `actual` and `desired` must match
xp_assert_close(actual, desired, check_0d=False) # no checks for 0-D vs scalar
```

**Pros:**
- clean solution, solves the problems from gh-21044

**Cons:**
- the keyword special-cases NumPy, hence the API isn't truly library-agnostic (but this is already the case before this PR)

*Remarks*:
- this seems like an obvious improvement 👍 the special-casing of NumPy seems to be necessary for us, however it does raise questions about how to generalise to a solution which would live upstream under the data-apis org *(would it be okay to treat NumPy as special or not?)*

## (2) Full module returning scalars

### Solution

```python
from scipy._lib._array_api_no_0d import xp_assert_close
...
xp_assert_close(actual, desired) # `actual` must be a scalar, regardless of the type of `desired`
xp_assert_close(actual, desired, check_0d=True) # `actual` must be 0-D if `desired` is 0-D
```

**Pros:**
- clean solution, doesn't increase verbosity while solving the problems from gh-21044

**Cons:**
- the "requiring scalar" semantics are abstracted away from the test functions to the import at the top of the module, so you have to scroll up to be aware of it

*Remarks*:
- I think the con is minor - it is already abstracted away before this PR, which is why some of the initial confusion arose

### Alternative (2A)

Require a keyword in every call to say that you expect a scalar. E.g.

```python
xp_assert_close(actual, desired, require_scalar=True)
```

**Pros:**
- more explicit and clear

**Cons:**
- very verbose when used in every call in a file

*Remarks*:
- I think that this is too verbose for the added clarity to be worth it

### Alternative (2B)

Cast `desired` to a scalar in every call via `[()]`.

**Pros:**
- more explicit (and clear?)

**Cons:**
- more verbose (unnecessarily?)
- See (3)

*Remarks*:
- I think that this is too verbose for any added clarity to be worth it

## (3) Module with some functions returning 0-D arrays, some returning scalars

### Solution

```python
from scipy._lib._array_api import xp_assert_close
...
xp_assert_close(actual, desired) # we want a 0-D array
xp_assert_close(actual, desired[()]) # we want a scalar
# XXX: would this be used?
xp_assert_close(actual, desired, check_0d=False) # we don't care either way
```

**Pros:**
- concise
- clear semantics for NumPy

**Cons:**
- while `[()]` works as a no-op for the alternative libraries which we test with[^1], it isn't standardised. See https://github.com/data-apis/array-api/issues/833. Hence this could be seen as semantically sketchy / a fudge for alternative backends rather than a watertight solution
- the `[()]` may be considered visually unappealing and less explicit/clear than a keyword

*Remarks:*

This seems like the best solution of the bunch to me, although I recognise that the problems described in the cons are real, hence it isn't an ideal solution

### Alternative (3A)

Like (2A), require a keyword for scalar checks:

```python
xp_assert_close(actual, desired) # we want a 0-D array
xp_assert_close(actual, desired, require_scalar=True) # we want a scalar
```

**Pros:**
- more explicit and clear

**Cons:**
- the special handling for NumPy arrays becomes built-in to the implementation of the assertion rather than handled by the user of the assertion at the call-point (or maybe this is a pro?)
- more verbose

*Remarks:*
- I would be happy with this alternative, even though it isn't my preference, if the concerns for the proposed solution here are deemed large

### Alternative (3B)

Just turn off the scalar/0-D check when a scalar is expected.

```python
xp_assert_close(actual, desired) # we want a 0-D array
xp_assert_close(actual, desired, check_0d=False) # we want a scalar
```

**Pros:**
- more explicit and clear

**Cons:**
- this is a regression in testing strictness - we no longer check that `actual` is a scalar, rather we just accept a scalar or a 0-D array

*Remarks:*
- this was the solution proposed in gh-21026. I think we want to avoid the regression in testing strictness, hence I didn't want to merge this solution

### Alternative (3C)

(3B) + do the scalar check separately. This could mean:

- a separate `assert_scalar_if_np` helper
- or, a separate test function in each instance of the problem (e.g. `test_func_returns_scalar`)

**Pros:**
- more explicit and clear
- avoids the regression for (3B)
- enables reduction of test redundancy, by only performing the scalar check once rather than in every use of the assertion
- a separate test is consistent with how we are checking behaviours for certain *inputs* elsewhere in the codebase, such as array-like input and empty array input

**Cons:**
- both options increase verbosity (the second especially would require a huge diff and a substantial amount of menial work to implement)
- this introduces the risk that performing the scalar check is forgotten about during development and thus gaps in the test coverage form - we no longer get the check "for free" just by using an assertion

*Remarks:*
- both cons seem quite serious for me, and I am of the opinion that they outweigh the pros here. But this is probably the most complex contest of ideas in this write-up

## (4) Module with functions varying behaviour based on `SCIPY_ARRAY_API`

### Solution

```python
from scipy._lib._array_api import xp_assert_close
...
actual = ndimage.sum_labels(input) # scalar without SCIPY_ARRAY_API, 0-D with SCIPY_ARRAY_API
xp_assert_close(actual, desired, check_0d=False) # just turn off the scalar/0-D check
```

**Pros:**
- simple
- does not reduce the strictness of the current tests - the scalar/0-D check was already turned off via `check_shape=False` in: https://github.com/scipy/scipy/blob/ada2bdb797b3e04d6b1f139971bca2612fd5bfce/scipy/_lib/_array_api.py#L390-L396

**Cons:**
- does not test for regressions (correctly returning scalar or 0-D based on if `SCIPY_ARRAY_API` is set)

*Remarks:*
- this seems fine since the check was turned off already, but I think it would be good to test for regressions in this situation too

 ### Alternative (4A)

Introduce a new kwarg to capture this behaviour, e.g.

```python
from scipy._lib._array_api import xp_assert_close
...
actual = ndimage.sum_labels(input) # scalar without SCIPY_ARRAY_API, 0-D with SCIPY_ARRAY_API
xp_assert_close(actual, desired, xp_now_0d=True) # check for 0-D with SCIPY_ARRAY_API, scalar without
```

**Pros:**
- introduces regression testing for this

**Cons:**
- I haven't thought of a good name for such a kwarg
- Makes the implementation a little more complex as it has to check the env var

*Remarks:*
- This seems like it would be an improvement to me, but could be saved for a follow-up

---

A follow-up to the PR could look at enabling some of the shape checks which are disabled across `ndimage`, now that instances (1), (3) and (4) are handled appropriately.

[^1]: it actually doesn't work for array-api-strict if a nonzero-dim array is passed. This can be easily worked around by guarding on `ndim == 0`, however. This is only a problem where a parameterised test results in a 0-dim array sometimes and a nonzero-dim array other times.

**Comment by h-vetinari** [comment]
I don't see why you're starting over after this RFC had found consensus. The only reason I'm leaving the implementation to you is because it's become too tiring to deal with your misplaced zeal in trying to combat all 0d arrays, everywhere. That doesn't mean the RFC restarts. You should implement it as agreed upon and fix the tests that got broken since I had finished my implementation.

You're also not explaining in your new list what problem your various alternatives are trying to solve. What are we evaluating against?

From my POV:
* No alternative using mass-deployment of `[()]` is acceptable
* No keyword-deviation from `check-*` is acceptable
* No sometimes-this-sometimes-that in the assert-functions
* No separate helpers beyond the two variants we've already discussed

In cases where we fail for numpy, but you absolutely require that other array-implementations avoid 0d-arrays, you can use `xp_assert_close(actual, desired, check_0d=not is_numpy(xp))`.

**Comment by lucascolley** [comment]
I plan to be at the community meeting this afternoon if anybody would like to discuss, by the way!

> In cases where we fail for numpy, but you absolutely require that other array-implementations avoid 0d-arrays, you can use xp_assert_close(actual, desired, check_0d=not is_numpy(xp)).

I think there is a misunderstanding here: array scalars are NumPy-specific. For all other backends we return 0-D arrays everywhere. We only need to be careful with array scalars for NumPy. `check_0d=not is_numpy(xp)` should be entirely equivalent to `check_0d=False`.

**Comment by lucascolley** [comment]
> The general xp_assert_close(np_array, np_array) must always pass, regardless of check_0d or not. You have the _no_0d variant to suppress this case.

Reading this, I can think of two alternatives which may meet your wishes for problem instance (3):

### Alternative (3D)

Use assertions from both `_array_api_no_0d` and `_array_api` in the same file:

```python
from scipy._lib._array_api import xp_assert_close
from scipy._lib._array_api_no_0d import xp_assert_close as xp_assert_close_no_0d
...
xp_assert_close(actual, desired) # we expect to return a 0d array
...
xp_assert_close_no_0d(actual, desired) # we expect to return a scalar 
```

### Alternative (3E)

Use `_array_api_no_0d` with `check_0d=True`:

```python
from scipy._lib._array_api_no_0d import xp_assert_close
...
xp_assert_close(actual, desired) # we expect to return a scalar
...
xp_assert_close(actual, desired, check_0d=True) # we expect to return a 0-D array
```

---

Would you support either of these alternatives?

**Comment by h-vetinari** [comment]
> Would you support either of these alternatives?

Sure, 3D sounds reasonable (pretty much exactly the cases we've designed for), _when warranted_. To me it should be exceptional though that both functions get included in a test module. Unless the split between the scalar/0d cases is close to even, IMO one should normally choose a sane default for the test module and then opt into the other behaviour where necessary.

That's why, if there's only a small handful of places in a test module that have an arrayness-accuracy in a given test module, I advocate for `xp_assert_close(actual, desired, check_0d=False)`. I think that's your 3B. In some other cases, it might end up being the other way around, i.e. 3E.

So IMO the rules could be summarized as:
1. If most `xp_assert_*` calls in a test module are arrayness-accurate, import `from scipy._lib._array_api`; opt out for exceptions with `check_0d=False` (3B)
2. if most `xp_assert_*` calls in a test module are arrayness-inaccurate (so we decide to live with that reality by enforcing no 0d results), then import `from scipy._lib._array_api_no_0d`; opt out for exceptions with `check_0d=True` (3E)
3. if `xp_assert_*` calls in a test module are near-evenly mixed between arrayness-accurate/inaccurate, import both variants (3D).

I think we could live with 1. & 2. without having 3., but I'm not against 3. per se.

**Comment by lucascolley** [comment]
> That's why, if there's only a small handful of places in a test module that have an arrayness-inaccuracy in a given test module, I advocate for xp_assert_close(actual, desired, check_0d=False). I think that's your 3B. In some other cases, it might end up being the other way around, i.e. 3E.

Thanks, this sounds closer to a solution that I would be okay with.

The main problem remaining for me is the asymmetry between (3B) and (3E) if we disallow `[()]` and the other alternatives like (3A) and (3C). For (3E), we can require either a scalar or a 0-D array:

```python
from scipy._lib._array_api_no_0d import xp_assert_close
...
xp_assert_close(actual, desired) # require scalar
...
xp_assert_close(actual, desired, check_0d=True) # require 0-D array (for 0-D `desired`)
```

However, for (3B), we are unable to require a scalar (without `[()]` or additional tools):

```python
from scipy._lib._array_api import xp_assert_close
...
xp_assert_close(actual, desired) # require 0-D array (for 0-D `desired`)
...
xp_assert_close(actual, desired, check_0d=False) # this does not require a scalar
xp_assert_close(actual, desired[()]) # the [()] is needed to require a scalar
```

> If most xp_assert_* calls in a test module are arrayness-accurate, import from scipy._lib._array_api; opt out for exceptions with check_0d=False (3B)

So that I understand, is your position that this reduced strictness is acceptable when most calls in a test module are arrayness-accurate, in virtue of allowing us to keep a more simple/clean/intuitive interface (no new keywords, no new helpers, no `[()]` etc.)?

What if a developer really values the scalar requirement, despite it only being applicable to a small proportion of the test module? Would you recommend that they split that part out into a separate file, so that they can use (3B) and (3E) independently?

**Comment by h-vetinari** [comment]
> So that I understand, is your position that this reduced strictness is acceptable when most calls in a test module are arrayness-accurate, in virtue of allowing us to keep a more simple/clean/intuitive interface (no extra imports, no new keywords, no new helpers, no `[()]` etc.)?

Yes.

> What if a developer really values the scalar requirement, despite it only being applicable to a small proportion of the test module?

I'd challenge that notion (outside of `stats`). No one "really" needs to check that the result is truly a scalar. It's a sanity-check in the face of an imperfect world, not some mission-critical aspect (_especially_ if the rest of the module isn't similarly affected). Doubly so because it's only about testing, not the implementation. And triply so because it doesn't affect all the other array backends.

> The main problem remaining for me is the asymmetry between [...]

The functions in `scipy._lib._array_api_no_0d` are the source of that asymmetry. You insist to ban 0d-arrays for `actual`, despite allowing them for `expected`. All other asymmetries stem from that.

> Thanks, this sounds closer to a solution that I would be okay with.

You really need to check your approach to collaboration here. You conceived of the current `xp_assert_*` functions as a "private testing utility" (this time with quote [source](https://github.com/scipy/scipy/pull/19989#issuecomment-2185266858)), suited to the needs of your [plural] main focus area. This whole RFC is the consequence of resolving the tension between some implementation choices that are unsuitable for wider use (e.g. #20959) and the need for the `xp_assert_*` infrastructure to serve all of SciPy.

"[...] a solution that I would be okay with" sounds as if you have a veto here, which you don't. You have one vote among many, and the majority of voices here coalesced around the proposal as written. If you want to write a competing proposal, go ahead, but don't change the outcome here after the fact.

You are also paying (seemingly) no heed to a multitude of increasingly bright red flags raised about this interaction by myself and other maintainers, e.g. @ev-br [stating](https://github.com/scipy/scipy/pull/21026#issuecomment-2297206585):
> While we're at it, I'll note that the whole array-api part of scipy is close to the least comfortable part of scipy to work on, by a long shot (and I'm comparing over multiple years of scipy contributions). Interactions here are way too nitpicky, bikesheddy, patronizing and argumentative.

I also find myself striking a much harsher tone than I'd like to (apologies...), because I cannot seem to get through to you that perfection in any given aspect is not just unattainable, but actively detrimental when taken in context of a larger whole. Especially for something so inconsequential as testing whether certain return values are 0d or scalar, the effort being expended here is cartoonishly out of proportion. I'll reiterate that I didn't want to get into this discussion (see opening paragraph of the OP). This came about because you [wanted](https://github.com/scipy/scipy/pull/19989#discussion_r1475911485) me to start using the `xp_assert_*` functions, which turned out to be clearly inadequate for the task.

So I've had my work on hold for half a year (over a harmless helper function in a single module), I spent a lot of effort coming up with a solution that fits all requirements *after* you had created _faits accomplis_ in the code base (IOW, cleaning up after a mess that's not mine), and after all that, I need to fight you tooth and nail for a handful of hypothetical inaccuracies in modules you aren't even a primary contributor to. It's a grotesque waste of time and energy for me, and I have just about no spoons left to keep contributing under these circumstances.

Your work, drive and contributions are very much appreciated, and your voice will not be ignored. Let me implore you though to leave some room for compromise (and in case this is not clear, the fact that you get these `_no_0d` functions to satisfy your use case is already a large concession[^1] vis-à-vis other legitimate goals for the codebase).

[^1]: not _my_ concession to grant (I also only have one vote of course), but acknowledging that there are other points of view and directions at odds with the `_no_0d` effort.
